### PR TITLE
PHP7 rewrite

### PR DESCRIPTION
**Pull request by esminis**
Rewrite to PHP7

### CODE REVIEW

**Comment by derickr** [comment]
A few things, this seems to be removing support for PHP 5 at the same time you're adding PHP 7... thta's not quite okay, as this is actually in use - and I'd need to be able to fix bugs that also go into the PHP 5 version.

You have added PHP/Zend Engine specific bits to code in lib/ , that's not okay as they should be usable without PHP as well. For example at https://github.com/derickr/quickhash/pull/7/files#diff-ed25556548acc44375d2182e66521efcR19 — I don't understand why though, as they don't seem to be used.

But the biggest issue with this right now, is that it removes PHP 5 support altogether.

File: qh_inthash.c [hunk 1]
```diff
@@ -297,12 +295,15 @@ static uint32_t qh_inthash_initialize_from_file(php_qh_inthash_obj *obj, php_str
...
 	char *filename;
-	int   filename_len;
+	size_t   filename_len;
 	long  size = 0, flags = 0;
 	php_stream *stream;
 
-	php_set_error_handling(EH_THROW, NULL TSRMLS_CC);
+    zend_error_handling error_handling;
+    zend_replace_error_handling(EH_THROW, NULL, &error_handling TSRMLS_CC);
+
```
Comment (derickr) [inline qh_inthash.c]: Please stick to the current CS, by using tabs for indentation.

File: qh_stringinthash.h [hunk 2]
```diff
@@ -25,8 +25,8 @@
 typedef struct _php_qh_stringinthash_obj php_qh_stringinthash_obj;
 
 struct _php_qh_stringinthash_obj {
-	zend_object   std;
 	qhi          *hash;
+	zend_object   std;
 };
```
Comment (derickr) [inline qh_stringinthash.h]: The PHP 5/PHP 7 supporting version could do something like:

Shared:
```
#if PHP_VERSION_ID < 70000
#define QH_PHP5_STD_OBJ zend_object   std;
#define QH_PHP7_STD_OBJ
#else
#define QH_PHP5_STD_OBJ
#define QH_PHP7_STD_OBJ zend_object   std;
#endif
```

Struct definition:
```
struct _php_qh_stringinthash_obj {
    QH_PHP5_STD_OBJ;
    qhi *hash
    QH_PHP7_STD_OBJ;
};
```

As an example.

File: qh_stringinthash.c [hunk 3]
```diff
@@ -521,15 +529,19 @@ PHP_METHOD(QuickHashStringIntHash, saveToString)
 	char              *string;
 	uint32_t           string_len;
 
-	php_set_error_handling(EH_THROW, NULL TSRMLS_CC);
```
Comment (derickr) [inline qh_stringinthash.c]: I think zend_replace_error_handling also works in PHP 5.4 and up, so I am okay with just replacing that (if it still works with 5.4 that is).

File: qh_stringinthash.c [hunk 4]
```diff
@@ -446,15 +455,11 @@ static int qh_stringinthash_string_validator(char *string, long length, uint32_t
...
 	uint32_t  hash_size;
-	uint32_t  string_store_size;
-	uint32_t  bucket_list_size;
 
 	if (string[0] != 'Q' || string[1] != 'H' || string[2] != 0x21) {
 		return 0;
 	}
 	hash_size = int_buffer[1];
-	string_store_size = int_buffer[2];
-	bucket_list_size = int_buffer[3];
```
Comment (derickr) [inline qh_stringinthash.c]: Why did you remove these?

File: qh_stringinthash.c [hunk 5]
```diff
@@ -121,52 +129,46 @@ void qh_register_class_stringinthash(TSRMLS_D)
...
-	memset(intern, 0, sizeof(php_qh_stringinthash_obj));
+	intern = ecalloc(1, sizeof(php_qh_stringinthash_obj) + zend_object_properties_size(class_type));
 	if (ptr) {
 		*ptr = intern;
 	}
 
 	zend_object_std_init(&intern->std, class_type TSRMLS_CC);
-#if PHP_MINOR_VERSION > 3
-	object_properties_init(&intern->std, class_type);
-#else
```
Comment (derickr) [inline qh_stringinthash.c]: As this is PHP 5.3 or later, I'm ok removing the #if def.

File: qh_stringinthash.c (hunk 4 above)
Comment (esminis) [inline qh_stringinthash.c]: Because they are not used in code

**Comment by esminis** [comment]
My target was to rewrite to support PHP7. I will modify code to support PHP5 but it will add lots of macros(thats the reason I did rewrite without PHP7 support). Will do this when I have free time, might weeks until then. Only stays tests that will not be compatible with PHP5

**Comment by esminis** [comment]
I modified lib because there were problems compiling, will check how to resolve them without adding includes

File: qh_stringinthash.c (hunk 3 above)
Comment (esminis) [inline qh_stringinthash.c]: Ok, will check

**Comment by Jan-E** [comment]
>My target was to rewrite to support PHP7. I will modify code to support PHP5 but it will add lots of macros(thats the reason I did rewrite without PHP5 support).

It is no ideal solution, but you can add the PHP5 and PHP7 sources in different directories and let the build system choose which directory it uses. See https://bitbucket.org/osmanov/pecl-event/src for an example of this approach.

**Comment by esminis** [comment]
Yea I know this solution but I doubt Derick will want this because it will create double code/double support. And double code I think is bad idea

**Comment by esminis** [comment]
One more problem I noticed PHP7 TSRM was changed, so this rewrite will not work good, so I am closing pull request for now, I will submit new pull request once I add support for PHP5 and fix TSRM
